src/librbd/Utils.h

```diff
@@ -59,6 +76,25 @@ class C_CallbackAdapter : public Context {
   }
 };
 
+template <typename T, void (T::*MF)(int), typename R>
+class C_RefCallbackAdapter : public Context {
+  R *refptr;
```

Comment:
hmm, this and the below change seems to be conflicting with the existing mock test structures since none of the tests currently inherit from `RefCountedObject`

Reply:
@MahatiC I haven't tried it, but in theory you should be able to use partial template specialization to handle that by just providing a `template<> class boost::intrusive_ptr<NameOfClassHere> { ... }`

Reply:
@dillaman Not sure if I'm missing something, this class expects a `boost::intrusive_ptr<RefCountedObj>`, so I think `MockTestClassName` will have to inherit from `RefCountedObj` to make it work, but we'll end up in the same situation for tests i.e. explicitly calling `put()` for tests as well.

Reply:
... but you can just use `ceph::ref_t<T>` instead of `RefCountedPtr` and use partial template specialization to override the case when it doesn't derive from `RefCountedObject`.

Alternatively, and perhaps cleaner in the unit test case, there really isn't a reason to instantiate `C_RefCallbackAdapter` if the classes don't inherit from `RefCountedObject`. Instead, the `create_context_callback` method would just use SFINAE to pick which version to use. If the `T` object derives from `RefCountedObject`, it would use your new `C_RefCallbackAdapter`, otherwise use the original.

Reply:
Done. Used the latter option, thanks.